### PR TITLE
codeintel: Protect indexes serving intel for branch and tag tips from eviction

### PR DESCRIPTION
**Pull request by efritz**
Previously, we tracked only the set of uploads that were visible _from the tip of the default branch_ for each repository. This PR expands the `lsif_uploads_visible_at_tip` table and commit graph calculation to also track the tip of _any_ non-stale branch and tag.

The `visible_at_tip` property in LSIF uploads and dumps still refers to the same concept (uploads that are visible from the tip of the default branch used in xrepo find references). All other uploads serving intelligence for the tip of another branch or tag are simply protected from being evicted based on age.

There is a maximum age threshold for branches and tags, after which the commit pointing to the branch head or tag is considered stale and will not be retained on the next code intel cleanup pass. These thresholds can be supplied globally as an env var (default is 3 months for branches, 1 year for tags). They can also be supplied via the database on a pre-repo basis (the table exists in this PR but it is not currently exposed for mutation via graphQL or the UI).

Partial effort towards #19120.

### CODE REVIEW

File: enterprise/internal/codeintel/stores/dbstore/commits.go [hunk 1]
```diff
@@ -269,6 +289,35 @@ const calculateVisibleUploadsDirtyRepositoryQuery = `
 UPDATE lsif_dirty_repositories SET update_token = GREATEST(update_token, %s), updated_at = %s WHERE repository_id = %s
 `
 
+// refineRetentionConfiguration returns the maximum age for no-stale branches and tags, effectively, as configured
+// for hte given repository. If there is no retention configuration for the given repository, the given default
```
Comment (Strum355) [inline enterprise/internal/codeintel/stores/dbstore/commits.go:293]:
```suggestion
// for the given repository. If there is no retention configuration for the given repository, the given default
```

File: enterprise/internal/codeintel/stores/dbstore/commits.go [hunk 2]
```diff
@@ -505,17 +556,32 @@ func (s *Store) persistUploadsVisibleAtTip(ctx context.Context, repositoryID int
...
 	vat.repository_id = %s AND
-	vat.upload_id NOT IN (SELECT source.upload_id FROM t_lsif_uploads_visible_at_tip source)
+	NOT EXISTS (
+		SELECT 1
+		FROM t_lsif_uploads_visible_at_tip source
+		WHERE
+			source.upload_id = vat.upload_id AND
+			source.branch_or_tag_name = vat.branch_or_tag_name AND
+			source.is_default_branch = vat.is_default_branch
+	)
```
Comment (Strum355) [inline enterprise/internal/codeintel/stores/dbstore/commits.go:584]: maybe a "quick win" to add `LIMIT 1` to the subqueries here?

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/commits.go:584]: Can do it for documentation, but the `NOT EXISTS (SELECT 1 ... )` should not pull additional rows than necessary in the query plan.

Comment (Strum355) [reply enterprise/internal/codeintel/stores/dbstore/commits.go:584]: ah good to know, can ignore then

**Review by Strum355 — APPROVED** [review]

**Pull request merged**